### PR TITLE
Make lazy freezing-aware.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/kotlin/konan/worker/Lazy.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                } else {
+                    do {
+                        result = valueFrozen_.get()
+                    } while (result === INITIALIZING)
+                }
+                assert(result !== UNINITIALIZED && result !== INITIALIZING)
+                @Suppress("UNCHECKED_CAST")
+                return result as T
+            } else {
+                if (result === UNINITIALIZED_VALUE) {
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/Lazy.kt]: What is `UNINITIALIZED_VALUE`? Where is it declared? How is it different from `UNINITIALIZED`?

File: runtime/src/main/kotlin/konan/worker/Lazy.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                    result = initializer_!!()
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/Lazy.kt]: What happens if `initializer_()`  freezes this object?

File: runtime/src/main/kotlin/konan/worker/Lazy.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            if (isFrozen) {
+                if (result !== UNINITIALIZED) {
+                    assert(result !== INITIALIZING)
+                    @Suppress("UNCHECKED_CAST")
+                    return result as T
+                }
+                // Barrier.
+                do {
+                    result = valueFrozen_.get()
+                } while (result === INITIALIZING)
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/Lazy.kt]: In case of recursive initialization we seem to get infinite loop here.

Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: Yes, as designed.

File: runtime/src/main/kotlin/konan/worker/Lazy.kt (hunk 1 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: Typo.

File: runtime/src/main/kotlin/konan/worker/Lazy.kt (hunk 2 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: Next statement will throw mutability exception. What do you suggest - go frozen path?

File: runtime/src/main/kotlin/konan/worker/Lazy.kt (hunk 3 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: Hung program is harder to debug than crashed one.

File: runtime/src/main/kotlin/konan/worker/Lazy.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: At least we can throw more understandable error for now.

Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: OK.

File: runtime/src/main/kotlin/konan/worker/Lazy.kt (hunk 3 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: I have quite an opposite opinion, if debugger is available. Anyway, what do you suggest?

Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/konan/worker/Lazy.kt]: I suggest to think a little about user-friendly handling for this case.

Btw, spin lock doesn't seem to be most suitable mutex implementation for this class. On the other hand, we are still missing proper lock implementations in our stdlib, which would be useful here and in many other cases.

**Pull request merged**